wire/merkle_tree.go

```diff
@@ -0,0 +1,310 @@
+
+	// Check merkle proofs for this hash.
+	for _, mp := range t.merkleProofs {
+		if mp.Index == -1 && mp.TxID != nil && mp.TxID.Equal(&hash) {
+			mp.Index = t.count
+			break
+		}
+	}
+
+	if len(t.layers) == 0 {
+		// First hash in tree
+		t.layers = []*merkleNodeLayer{newMerkleNodeLayer(hash)}
+		t.count = 1
+		return
+	}
+
+	next := hash
+	t.count++
+
+	// Calculate a new hash up the tree
+	s := sha256.New()
+	for _, layer := range t.layers {
+		// Append to this row
+		layer.addHash(next)
+
+		l := layer.len()
+		if l%2 != 0 {
+			return // Above layers do not need to be updated.
+		}
+
+		left := layer.nextLastHash()
+		right := next
+
+		// Even number of hashes in layer. Hash last 2 hashes together to add to layer above.
+		s.Write(left[:])
+		s.Write(right[:])
+		b := s.Sum(nil)
+		s.Reset()
+		next = bitcoin.Hash32(sha256.Sum256(b)) // Sum again for double SHA256
+		if t.prune {
+			layer.clear() // Clear out hashes that aren't needed anymore
+		}
+
+		// Check active merkle proofs for next element in path
+		for _, mp := range t.merkleProofs {
+			if mp.Index == -1 {
+				continue // txid not found yet
+			}
+
+			if mp.root.Equal(&left) {
+				mp.AddHash(right, next)
+			} else if mp.root.Equal(&right) {
+				mp.AddHash(left, next)
+			}
+		}
+	}
+
+	// Append new layer
+	t.layers = append(t.layers, newMerkleNodeLayer(next))
+}
+
+func (t *MerkleTree) RootHash() bitcoin.Hash32 {
+	if t.count == 0 {
+		return bitcoin.Hash32{} // zero hash should never be valid
+	}
+	if t.count == 1 {
+		return t.layers[0].lastHash()
+	}
+
+	// Check for odd length layer to calculate up from
+	var next *bitcoin.Hash32
+	s := sha256.New()
+	for d, layer := range t.layers {
+		l := layer.len()
+
+		if next != nil {
+			// Odd layer was below this. So keep calculating up.
+			if l%2 == 0 {
+				// Layer will be odd length with new hash, so hash next hash with itself
+				s.Write(next[:])
+				s.Write(next[:])
+				b := s.Sum(nil)
+				s.Reset()
+				h := bitcoin.Hash32(sha256.Sum256(b)) // Sum again for double SHA256
+				next = &h
+				continue
+			}
+
+			// hash last hash with next hash
+			s.Write(layer.lastBytes())
+			s.Write(next[:])
+			b := s.Sum(nil)
+			s.Reset()
+			h := bitcoin.Hash32(sha256.Sum256(b)) // Sum again for double SHA256
+			next = &h
+			continue
+		}
+
+		if l%2 != 0 {
+			if l == 1 && d == len(t.layers)-1 {
+				// Last layer so this is the root hash
+				return layer.lastHash()
+			}
+
+			// Odd length layer. Calculate from here up.
+			s.Write(layer.lastBytes())
+			s.Write(layer.lastBytes())
+			b := s.Sum(nil)
+			s.Reset()
+			h := bitcoin.Hash32(sha256.Sum256(b)) // Sum again for double SHA256
+			next = &h
+		}
+	}
+
+	if next == nil {
+		return bitcoin.Hash32{} // zero hash should never be valid
+	}
+	return *next
+}
+
+// FinalizeMerkleProofs finalizes the merkle proofs and returns the root hash and the merkle proofs.
+func (t MerkleTree) FinalizeMerkleProofs() (bitcoin.Hash32, []*MerkleProof) {
```

Comment:
SonarCloud gives this func 52 out 15 allowed.

The loop is long, and deep. Please consider ways to simplify.

Reply:
I was able to move the hash calculation block of code out to another function with less than a 1% performance hit.